File: src/main/java/io/github/m/Either.java
package io.github.m;

/**
 * M wrapper class for either.
 */
public interface Either extends Value {
    /**
     * The value of this either.
     */
    Value value();

    /**
     * Returns left if this either is left, or right if this either is right.
     */
    Value fold(Value left, Value right);

    /**
     * Partially applies this value to an argument.
     */
    @Override
    default Value invoke(Value arg) {
        return new Partial(this, arg);
    }

    /**
     * Alias of fold.
     */
    @Override
    default Value invoke(Value arg1, Value arg2) {
        return fold(arg1, arg2).invoke(value());
    }

    /**
     * The left instance of either such that fold(x, y) == x.
     */
    final class Left implements Either {
        private final Value value;

        public Left(Value value) {
            this.value = value;
        }

        @Override
        public String toString() {
            return "Left(" + value + ")";
        }

        @Override
        public Value value() {
            return value;
        }

        @Override
        public Value fold(Value left, Value right) {
            return left;
        }
    }

    /**
     * The right instance of either such that fold(x, y) == y.
     */
    final class Right implements Either {
        private final Value value;

        public Right(Value value) {
            this.value = value;
        }

        @Override
        public String toString() {
            return "Right(" + value + ")";
        }

        @Override
        public Value value() {
            return value;
        }

        @Override
        public Value fold(Value left, Value right) {
            return right;
        }
    }

    /**
     * Converts a value to an either.
     */
    static Either from(Value value) {
        if (value instanceof Either) {
            return (Either) value;
        } else {
            return (Either) value.invoke(left, right);
        }
    }

    @MField(name = "left")
    Value left = new Value.Impl1("left", Left::new);

    @MField(name = "right")
    Value right = new Value.Impl1("right", Right::new);
}

File: src/main/kotlin/io/github/m/Either.kt
package io.github.m

/**
 * M wrapper class for either.
 */
sealed class Either : Value {
    /**
     * The left value of either.
     */
    data class Left(val value: Value) : Either() {
        override fun invoke(arg: Value) = Value { arg2 -> this(arg, arg2) }
        override fun invoke(arg1: Value, arg2: Value) = arg1(value)
    }

    /**
     * The right value of either.
     */
    data class Right(val value: Value) : Either() {
        override fun invoke(arg: Value) = Value { arg2 -> this(arg, arg2) }
        override fun invoke(arg1: Value, arg2: Value) = arg2(value)
    }

    companion object {
        fun from(value: Value) = value as? Either ?: value(Value(::Left), Value(::Right)) as Either
    }

    /**
     * M either definitions.
     */
    @Suppress("unused")
    object Definitions {
        @MField(name = "left")
        @JvmField
        val left: Value = Value.Impl1 { x -> Left(x) }

        @MField(name = "right")
        @JvmField
        val right: Value = Value.Impl1 { x -> Right(x) }

        @MField(name = "left?")
        @JvmField
        val isLeft: Value = Value.Impl1 { x -> Bool(from(x) is Left) }

        @MField(name = "right?")
        @JvmField
        val isRight: Value = Value.Impl1 { x -> Bool(from(x) is Right) }
    }
}